controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -331,24 +333,113 @@ private ReaderGroupConfiguration getRGConfigurationFromRecord(final String scope
                   return CompletableFuture.completedFuture(CreateReaderGroupStatus.Status.SCOPE_NOT_FOUND);
          }
          //2. check state of the ReaderGroup, if found
-         return Futures.exceptionallyExpecting(streamMetadataStore.getReaderGroupState(scope, rgName, true, null, executor),
-                e -> Exceptions.unwrap(e) instanceof StoreException.DataNotFoundException, ReaderGroupState.UNKNOWN)
-                .thenCompose(state -> {
-                if (state.equals(ReaderGroupState.UNKNOWN) || state.equals(ReaderGroupState.CREATING)) {
-                   CreateReaderGroupEvent event = new CreateReaderGroupEvent(scope, rgName, requestId);
-                   //3. Create Reader Group Metadata and submit event
-                   return eventHelper.addIndexAndSubmitTask(event,
-                              () -> streamMetadataStore.createReaderGroup(scope, rgName, config, createTimestamp, null, executor))
-                              .thenCompose(x -> eventHelper.checkDone(() -> isRGCreated(scope, rgName, executor))
-                              .thenApply(done -> CreateReaderGroupStatus.Status.SUCCESS));
-                }
-                // idempotent call
-                return CompletableFuture.completedFuture(CreateReaderGroupStatus.Status.SUCCESS);
-             });
+         return isRGCreationComplete(scope, rgName)
+                 .thenCompose(complete -> {
+                     if (!complete) {
+                         Map<String, RGStreamCutRecord> startStreamCuts = config.getStartingStreamCuts().entrySet().stream()
+                                 .collect(Collectors.toMap(e -> e.getKey().getScopedName(),
+                                         e -> new RGStreamCutRecord(ImmutableMap.copyOf(ModelHelper.getStreamCutMap(e.getValue())))));
```

Comment:
same here and everywhere else where this method is called. we are creating copies that could be avoided if the method returned an immutable map itself

Reply:
Same as earlier.

Reply:
done in PR5466